### PR TITLE
Vision and damage cheat

### PR DESCRIPTION
**Pull request by Robi321**

### CODE REVIEW

**Comment by Jaccosf** [comment]
@Robi321 The events in the client you've added are (still) named AntiSalt.

File: ac_c.lua [hunk 1]
```diff
@@ -54,6 +54,27 @@ Citizen.CreateThread(function()
 	end
 end)
 
+Citizen.CreateThread(function()
+	while true do
+		Citizen.Wait(60000)
+		if NetworkIsInSpectatorMode() then
+			TriggerServerEvent("AntiCheese:Misc", "Cheating", "Spectator", true)
```
Comment (Blumlaut) [inline ac_c.lua:61]: I would recommend making seperate event for these as you are reporting what exactly it does anyway, having them grouped into one "generic" event just makes them harder to customize.

File: ac_s.lua [hunk 2]
```diff
@@ -851,6 +865,22 @@ Citizen.CreateThread(function()
 		end
 	end)
 
+	AddEventHandler("clearPedTasksEvent", function(source, data)
```
Comment (Blumlaut) [inline ac_s.lua:868]: wouldnt this also trigger if tasks of normal peds were cleared by a script? I would narrow it down and double check if the source of the event is not also the ped its being executed on, and making sure the ped is a player.

File: ac_s.lua [hunk 3]
```diff
@@ -851,6 +865,22 @@ Citizen.CreateThread(function()
...
+		if data.immediately then
+			CancelEvent()
+			if Components["client.cleartask"] and not IsPlayerAceAllowed(source,"anticheese.bypass") then
+				local license, steam = GetPlayerNeededIdentifiers(source)
+				local name = GetPlayerName(source)
+				if not extrainfo then extrainfo = "no extra informations provided" end
+				local isKnown, isKnownCount, isKnownExtraText, alreadyBanned = WarnPlayer(source,reason, banInstantly)
+	
+				if not alreadyBanned then
+					SendWebhookMessage(webhook,"**"..reason.."** \n```\nUser:"..name.."\n"..license.."\n"..steam.."\n"..extrainfo.."\nAnticheat Flags:"..isKnownCount..""..isKnownExtraText.." ```")
```
Comment (Blumlaut) [inline ac_s.lua:878]: reason is missing

File: ac_config.lua [hunk 4]
```diff
@@ -7,6 +7,8 @@ Components = {
 	["client.superjump"] = false, -- protect against super jump cheats
 	["client.weaponblacklist"] = false, -- blacklist certain weapons in ac_c.lua
 	["client.carblacklist"] = false, -- blacklist certain cars in ac_c.lua
+	["client.misc"] = true, -- anti night, thermal vision, anti multi damage and spectate
+	["server.cleartask"] = true, -- detect kicking from car
```
Comment (Blumlaut) [inline ac_config.lua:11]: i would further elaborate on the comment, something like `detects if a player clears tasks of another player, e.g. kicks them from a car`

File: ac_config.lua [hunk 5]
```diff
@@ -7,6 +7,8 @@ Components = {
 	["client.superjump"] = false, -- protect against super jump cheats
 	["client.weaponblacklist"] = false, -- blacklist certain weapons in ac_c.lua
 	["client.carblacklist"] = false, -- blacklist certain cars in ac_c.lua
+	["client.misc"] = true, -- anti night, thermal vision, anti multi damage and spectate
```
Comment (Blumlaut) [inline ac_config.lua:10]: misc is awfully generic, i'd make them seperate detections.

**Review by Blumlaut — CHANGES_REQUESTED** [review]

**Comment by Robi321** [comment]
Fixed

**Comment by Blumlaut** [comment]
My other concern regarding ClearPedTasks has not been addressed yet
https://github.com/Blumlaut/anticheese-anticheat/pull/16#discussion_r695089766

File: ac_s.lua [hunk 6]
```diff
@@ -850,6 +906,23 @@ Citizen.CreateThread(function()
 			end
 		end
 	end)
+		
+	AddEventHandler("clearPedTasksEvent", function(ped, source, data)
+    		if data.immediately then
+        		if IsPedAPlayer((ped)) then
```
Comment (Blumlaut) [inline ac_s.lua:912]: wouldnt this trigger if i call Clear Tasks on myself? or does the event not take that into account

Comment (Robi321) [reply ac_s.lua]: if think now what i have done is good to use

**Pull request merged**